migrations/sql/V2019.04.25.11.30__add_variance_applications_metadata.sql

```diff
@@ -0,0 +1,105 @@
+    ('NAP', 'Not Applicable', 'system-mds', 'system-mds'),
+    ('APP', 'Approved', 'system-mds', 'system-mds'),
+    ('DEN', 'Denied', 'system-mds', 'system-mds');
+
+
+ALTER TABLE variance
+ADD COLUMN variance_application_status_code character varying(3) DEFAULT 'REV' NOT NULL,
+ADD COLUMN ohsc_ind                         boolean              DEFAULT FALSE NOT NULL,
+ADD COLUMN union_ind                        boolean              DEFAULT FALSE NOT NULL,
+ADD COLUMN inspector_id                     integer                                    ,
+ADD COLUMN variance_no                      integer                            NOT NULL;
+
+COMMENT ON COLUMN variance.variance_no IS 'A public, unique identifier for a mine variance.';
+
+ALTER TABLE variance
+ALTER COLUMN issue_date  DROP NOT NULL,
+ALTER COLUMN expiry_date DROP NOT NULL;
+
+ALTER TABLE variance
+ALTER COLUMN issue_date    DROP DEFAULT,
+ALTER COLUMN expiry_date   DROP DEFAULT,
+ALTER COLUMN received_date DROP DEFAULT;
+
+ALTER TABLE variance
+ADD UNIQUE (mine_guid, variance_no);
+
+ALTER TABLE variance
+ADD CONSTRAINT variance_status_variance_application_status_code_fkey
+    FOREIGN KEY (variance_application_status_code)
+    REFERENCES variance_application_status_code(variance_application_status_code)
+    DEFERRABLE INITIALLY DEFERRED,
+ADD CONSTRAINT variance_inspector_core_user_fkey
+    FOREIGN KEY (inspector_id) REFERENCES core_user(core_user_id) DEFERRABLE INITIALLY DEFERRED,
+
+-- Business Rule Constraints
+
+  -- inspector_id
+ADD CONSTRAINT inspector_required_on_reviewed_application
+    CHECK (NOT (inspector_id IS NULL AND (
+              variance_application_status_code = 'APP' OR variance_application_status_code = 'DEN'
+          ) ) ),
+
+  -- issue_date
+ADD CONSTRAINT issue_date_required_on_approved_variance_else_forbidden
+    CHECK (
+        (NOT (issue_date IS NULL AND variance_application_status_code = 'APP') )
+        AND
+        (NOT (issue_date IS NOT NULL AND variance_application_status_code != 'APP') )
+    ),
+
+  -- expiry_date
+ADD CONSTRAINT expiry_date_required_on_approved_variance_else_forbidden
+    CHECK (
+        (NOT (expiry_date IS NULL AND variance_application_status_code = 'APP') )
+        AND
+        (NOT (expiry_date IS NOT NULL AND variance_application_status_code != 'APP') )
+    );
+
+CREATE OR REPLACE FUNCTION add_incremented_variance_no() RETURNS trigger AS $$
+    DECLARE
+        max_mine_variance_no integer;
+    BEGIN
+        SELECT
+            CASE
+                WHEN (MAX(variance_no) IS NOT NULL)
+                THEN MAX(variance_no)
+                ELSE 0
+            END
+            FROM variance
+            WHERE mine_guid = NEW.mine_guid INTO max_mine_variance_no;
+
+        NEW.variance_no := max_mine_variance_no + 1;
+        RETURN NEW;
+    END;
+$$LANGUAGE plpgsql;
+
+CREATE TRIGGER new_variance_no
+  BEFORE INSERT
+  ON variance
+  FOR EACH ROW
+  EXECUTE PROCEDURE add_incremented_variance_no();
```

Comment:
Is all of this just for an autoincrementing column. i think we can do this as a sequence, just as it was done for serial id's.

Reply:
It increments on a per-mine basis. It's not a global sequence

Reply:
It was just explained to be incrementing per mine. Which was a requirement asked for mine-incidents, but we asked if a globally incrementing integer was fine and Aaron was fine with that.  Not sure if this is a different case but is feels like a similar situation.